Test HTMLMediaElement preload 'none' is ignored for MediaStream as src or srcObject

wolenetz opened the pull request:
@foolip This is the MediaStream side of the change mentioned in Chromium https://codereview.chromium.org/1881733004/. I expect to make a similar PR for the media-source (and bogus blob URL) portion of tests contained within that Chromium CL shortly.

Regarding these MediaStream tests, I confirmed that Chrome ToT fails prior to https://codereview.chromium.org/1881733004/ and passes after. The fails are actually timeouts though (and the console shows 2 exceptions: the unexpected 'suspend' event occurs, and the networkState is IDLE, not LOADING after 1 second. Do you have a better mechansim to suggest other than using setTimeout() to help prevent slow machine flakiness and enable fast test runs? I couldn't think of one immediately that would be interoperable, hence the 1 second setTimeout() used in these new tests.

wolenetz committed:
Test HTMLMediaElement preload 'none' is ignored for MediaStream as src and srcObject

foolip commented on mediacapture-streams/MediaStream-MediaElement-preload-none-src.html:
```diff
@@ -0,0 +1,38 @@
+        <h1 class="instructions">Description</h1>
+        <p class="instructions">This test checks that the HTMLMediaElement preload 'none' attribute value is ignored for MediaStream object URLs used as src.</p>
+
+        <audio id="aud"></audio>
```
You can drop the id and use `document.querySelector("audio")`. You can also use put `preload=none` here in the markup.

wolenetz replied:
Thanks. I'll have these fixed later in this PR.

foolip commented on mediacapture-streams/MediaStream-MediaElement-preload-none-src.html:
```diff
@@ -0,0 +1,38 @@
+        <script>
+            var aud = document.getElementById("aud");
+            aud.preload="none";
+            var t = async_test("Tests that the HTMLMediaElement preload attribute value 'none' is ignored for MediaStream object URLs used as src.");
```
The default title is taken from `document.title`, so this can be simplified to `async_test(function(t) { ... } )`

wolenetz replied:
I didn't know that. Thanks. I'll have these fixed later in this PR.

foolip commented on mediacapture-streams/MediaStream-MediaElement-preload-none-src.html:
```diff
@@ -0,0 +1,38 @@
+                    // Let the media element load algorithm proceed, ensure no 'suspend' is fired, and verify that the
+                    // networkState indicates we are loading and not idle. Give it a second.
+                    setTimeout(function() { assert_equals(aud.networkState, 2); t.done(); }, 1000);
+                }), function(error) {});
```
t.unreached_func() will make the test fail fast if the error callback is invoked.

wolenetz replied:
Good point. I'll keep the wrapping inside function(error) {...} though since I'm not sure |error| is meaningful as argument to unreached_func. I'll have this fixed later in this PR.

wolenetz replied:
Note that even t.unreached_func() still results in Timeout (and doesn't fail fast).

foolip commented on mediacapture-streams/MediaStream-MediaElement-preload-none-src.html:
```diff
@@ -0,0 +1,38 @@
+                navigator.getUserMedia({audio:true}, t.step_func(function(stream)
+                {
+                    // The optional deferred load steps (for preload none) for MediaStream resources should be skipped.
+                    aud.addEventListener("suspend", function() { assert_unreached("'suspend' should not occur here."); }, false);
```
`t.unreached_func("'suspend' should not occur here.")` and the 3rd argument isn't needed, false is the default.

wolenetz replied:
Thanks. I'll have these fixed later in this PR.

foolip commented on mediacapture-streams/MediaStream-MediaElement-preload-none-src.html:
```diff
@@ -0,0 +1,38 @@
+                {
+                    // The optional deferred load steps (for preload none) for MediaStream resources should be skipped.
+                    aud.addEventListener("suspend", function() { assert_unreached("'suspend' should not occur here."); }, false);
+                    aud.src = URL.createObjectURL(stream);
```
Can you `t.add_cleanup(function() { URL.revokeObjectURL(aud.src); })`?

wolenetz replied:
Good point - this demonstrates best practice. I'll have that later in this PR.

foolip commented on mediacapture-streams/MediaStream-MediaElement-preload-none-src.html:
```diff
@@ -0,0 +1,38 @@
+                    // The optional deferred load steps (for preload none) for MediaStream resources should be skipped.
+                    aud.addEventListener("suspend", function() { assert_unreached("'suspend' should not occur here."); }, false);
+                    aud.src = URL.createObjectURL(stream);
+                    assert_equals(aud.networkState, 3); // Resource selection is active.
```
Suggest `aud.NETWORK_NO_SOURCE`, at least I never remember these numbers.

wolenetz replied:
Excellent point. I'll have that later in this PR.

foolip commented on mediacapture-streams/MediaStream-MediaElement-preload-none-src.html:
```diff
@@ -0,0 +1,38 @@
+
+                    // Let the media element load algorithm proceed, ensure no 'suspend' is fired, and verify that the
+                    // networkState indicates we are loading and not idle. Give it a second.
+                    setTimeout(function() { assert_equals(aud.networkState, 2); t.done(); }, 1000);
```
Callbacks should be wrapped in `t.step_func()`, but maybe there's another way. If the load isn't suspended, `readyState` should reach at the very least HAVE_CURRENT_DATA, so could you instead wait for the `loadeddata` event? If that event isn't fired, I wouldn't be totally surprised, but it would be a bug.

wolenetz replied:
'loadeddata': excellent point. I'll use that instead of setTimeout.

foolip commented on mediacapture-streams/MediaStream-MediaElement-preload-none-src.html:
```diff
@@ -0,0 +1,38 @@
+<!DOCTYPE html>
+<!-- Copyright © 2016 Chromium authors and World Wide Web Consortium, (Massachusetts Institute of Technology, ERCIM, Keio University, Beihang). -->
```
I don't see this in any other files, were you asked to add it? I like minimalism and git log for figuring out where things came from :)

wolenetz replied:
Yes - per separate discussion with w3c folks and internal Google, upstreamed versions of Chromium blink layout tests like this one will need this stamp. I too like minimalism FWIW :)

wolenetz replied:
For this particular PR though, it's less clear to me that this stamp is needed since these are direct-to-wpt, though since authored by a Googler (and reviewed by w3c members), this stamp is still correct and perhaps best to keep even in this case.

foolip commented on mediacapture-streams/MediaStream-MediaElement-preload-none-srcObject.html:
```diff
@@ -0,0 +1,38 @@
+        <script>
+            var aud = document.getElementById("aud");
+            aud.preload="none";
+            var t = async_test("Tests that the HTMLMediaElement preload attribute value 'none' is ignored for MediaStream used as srcObject.");
```
Much of this is the same, can you put the framework in a js file and pass in a callback that will either set srcObject or the src attribute?

wolenetz replied:
I agree lots of duplication. Rather than 3 files (2 html + 1 js), I'll merge it all into one html and refactor out a utility test function in that same html. If later we find that mediacapture-streams needs that utility elsewhere, a separate js could be used then.

wolenetz commented:
@foolip I believe I have addressed all of your comments. Note that test.unreached_func(...) does not immediately stop the test. These tests timeout on failure still, even if the 'suspend' event is handled by testPreloadNone(), which uses unreached_func().

wolenetz commented on mediacapture-streams/MediaStream-MediaElement-preload-none.html:
```diff
@@ -0,0 +1,61 @@
+                // The optional deferred load steps (for preload none) for MediaStream resources should be skipped.
+                mediaElement.addEventListener("suspend", t.step_func(function()
+                {
+                    t.unreached_func("'suspend' should not occur here.");
```
Oops - I need to *execute* this function to get the test to fail fast here.

wolenetz commented on mediacapture-streams/MediaStream-MediaElement-preload-none.html:
```diff
@@ -0,0 +1,61 @@
+                        aud.src = URL.createObjectURL(stream);
+                        t.add_cleanup(function() { URL.revokeObjectURL(aud.src); });
+                    }));
+                }), function(error) { t.unreached_func(); });
```
ditto

wolenetz commented on mediacapture-streams/MediaStream-MediaElement-preload-none.html:
```diff
@@ -0,0 +1,61 @@
+                navigator.getUserMedia({video:true}, t.step_func(function(stream)
+                {
+                    testPreloadNone(t, vid, t.step_func(function() { vid.srcObject = stream; }));
+                }), function(error) { t.unreached_func(); });
```
ditto

wolenetz commented:
@foolip OK. I think this is now ready again for your review.

wolenetz commented on mediacapture-streams/MediaStream-MediaElement-preload-none.html:
```diff
@@ -0,0 +1,61 @@
+                // The optional deferred load steps (for preload none) for MediaStream resources should be skipped.
+                mediaElement.addEventListener("suspend", t.step_func(function()
+                {
+                    t.unreached_func("'suspend' should not occur here.")();
```
Would it be better to just assert_unreached(...) here?

wolenetz replied:
Ohh! I see now. Instead of t.step_func... just t_unreached_func. I'll update this again shortly.

wolenetz commented:
Ready for review again, finally (I hope :) ).

foolip commented on mediacapture-streams/MediaStream-MediaElement-preload-none.html:
```diff
@@ -0,0 +1,58 @@
+            function testPreloadNone(t, mediaElement, setSourceStreamFunc)
+            {
+                // The optional deferred load steps (for preload none) for MediaStream resources should be skipped.
+                mediaElement.addEventListener("suspend", t.unreached_func("'suspend' should not occur here."));
```
I guess I should ask for "should not be fired" here as well for consistency, even though I didn't complain the first time around :)

wolenetz replied:
done :)

foolip commented on mediacapture-streams/MediaStream-MediaElement-preload-none.html:
```diff
@@ -0,0 +1,58 @@
+                        aud.src = URL.createObjectURL(stream);
+                        t.add_cleanup(function() { URL.revokeObjectURL(aud.src); });
+                    }));
+                }), function(error) { t.unreached_func()(); });
```
Hmm, this still looks strange. Since you don't use the error at all, does a plain `t.unreached_func("getUserMedia error callback was invoked")` or similar not do the trick?

foolip replied:
If you do use the error in some way, then `assert_unreached('foo ' + error)` should do the trick instead of `t.unreached_func()()`.

wolenetz replied:
SGTM. In this context, if I "block", not "allow" the getUserMedia request, |error| in string form is "object NavigatorUserMediaError" -- not really useful. I'll go with t.unreached_func(...). If a similar test wanted to use error, I'm not clear that assert_unreached would be right in a callback -- it would need to be contained within a test.step_func, right?

foolip replied:
Yeah, you would have to do `t.step_func(function(error) { assert_unreached('foo ' + error); })` I think, or similar.

foolip commented on mediacapture-streams/MediaStream-MediaElement-preload-none.html:
```diff
@@ -0,0 +1,58 @@
+                navigator.getUserMedia({video:true}, t.step_func(function(stream)
+                {
+                    testPreloadNone(t, vid, t.step_func(function() { vid.srcObject = stream; }));
+                }), function(error) { t.unreached_func()(); });
```
Ditto.

wolenetz replied:
Done.

foolip commented:
LGTM % small nits, let me know when ready to merge!

wolenetz commented:
@foolip I believe this is ready to merge.
